### PR TITLE
[src] Add support for generating classes that throw PlatformNotSupportedExceptions in the Xamarin.iOS.dll implementation assembly for Mac Catalyst. Fixes #10213.

### LINKED ISSUE #10213: Mac Catalyst: Complete the type forwarding assembly (Xamarin.iOS.dll) to throw PNSE for types not in Mac Catalyst

### PR DESCRIPTION
**Pull request by rolfbjarne**
This adds any types in the Xamarin.iOS version of Xamarin.iOS.dll that are not
available in Xamarin.MacCatalyst.dll to the Xamarin.iOS.dll implementation
assembly we have for Mac Catalyst. The implementation of these types will
throw PlatformNotSupportedExceptions.

This means that the linker will not run into any missing API when linking Mac
Catalyst apps, which is necessary for our linksdk and linkall test suites,
because those test suites use MonoTouch.Dialog, of which there is no Mac
Catalyst version (yet at least).

Fixes https://github.com/xamarin/xamarin-macios/issues/10213.

### CODE REVIEW

File: src/generate-type-forwarders/Program.cs [hunk 1]
```diff
@@ -15,10 +27,468 @@ public static int Main (string [] args)
...
+				if (!string.IsNullOrEmpty (type.Namespace)) {
+					sb.Append (type.Namespace);
+					sb.Append ('.');
+				}
+			}
+
+			var arrayType = type as ArrayType;
+			if (arrayType != null) {
+				sb.Append (arrayType.ElementType.Name);
+				sb.Append ("[");
```
Comment (spouliot) [inline src/generate-type-forwarders/Program.cs:50]: minor: `'['` is (or used to be) more efficient

File: src/generate-type-forwarders/Program.cs [hunk 2]
```diff
@@ -15,10 +27,468 @@ public static int Main (string [] args)
...
+				sb.Append (new string (',', arrayType.Rank - 1));
+				sb.Append ("]");
```
Comment (spouliot) [inline src/generate-type-forwarders/Program.cs:52]: minor: `']'` is (or used to be) more efficient

File: src/generate-type-forwarders/Program.cs [hunk 3]
```diff
@@ -15,10 +27,468 @@ public static int Main (string [] args)
...
+			EmitParameters (sb, method.Parameters);
+			sb.Append (")");
+			if (constraints != null)
+				sb.Append (constraints);
+			if (method.IsAbstract) {
+				sb.AppendLine (";");
+			} else {
+				sb.AppendLine ();
+				// if there is a base constructor with the same signature, call it
+				if (method.IsConstructor && method.Parameters.Count > 0) {
```
Comment (spouliot) [inline src/generate-type-forwarders/Program.cs:201]: `method.HasParameters` will avoid allocation of the collection from Cecil

File: src/generate-type-forwarders/Program.cs [hunk 4]
```diff
@@ -15,10 +27,468 @@ public static int Main (string [] args)
...
+				v.Parameters.Count == method.Parameters.Count &&
+				v.GenericParameters.Count == method.GenericParameters.Count
+			), method.Parameters, out matchingMethod);
+		}
+
+		static bool AnyParameterTypeMatch (IEnumerable<MethodDefinition> methods, IList<ParameterDefinition> parameters, out MethodDefinition matchingMethod)
+		{
+			matchingMethod = null;
+
+			foreach (var method in methods) {
```
Comment (spouliot) [inline src/generate-type-forwarders/Program.cs:250]: `HasParameters`

File: src/generate-type-forwarders/Program.cs [hunk 5]
```diff
@@ -15,10 +27,468 @@ public static int Main (string [] args)
...
+
+		static void EmitField (StringBuilder sb, FieldDefinition fd, int indent)
+		{
+			var strIndent = new string ('\t', indent);
+			if (fd.DeclaringType.IsEnum) {
+				if (!fd.IsStatic)
+					return;
+				sb.Append (strIndent);
+				sb.Append (fd.Name);
+				sb.AppendLine (",");
```
Comment (spouliot) [inline src/generate-type-forwarders/Program.cs:277]: `','`

File: src/generate-type-forwarders/Program.cs [hunk 6]
```diff
@@ -15,10 +27,468 @@ public static int Main (string [] args)
...
+			} else {
+				sb.Append (strIndent);
+				sb.Append ("public ");
+				if (fd.IsStatic)
+					sb.Append ("static ");
+				EmitTypeName (sb, fd.FieldType);
+				sb.Append (" ");
+				sb.Append (fd.Name);
+				sb.AppendLine (";");
```
Comment (spouliot) [inline src/generate-type-forwarders/Program.cs:286]: `';'`

File: src/generate-type-forwarders/Program.cs [hunk 7]
```diff
@@ -15,10 +27,468 @@ public static int Main (string [] args)
...
+		static void EmitPNSE (StringBuilder sb, TypeDefinition type, int indent)
+		{
+			if (!type.IsNested)
+				sb.AppendLine ($"namespace {type.Namespace} {{");
+			var strIndent = new string ('\t', indent);
+			if (type.BaseType?.FullName == "System.MulticastDelegate") {
+				sb.Append ($"{strIndent}public delegate ");
+				var invoke = type.Methods.First (v => v.Name == "Invoke");
+				EmitTypeName (sb, invoke.ReturnType);
+				sb.Append (" ");
```
Comment (spouliot) [inline src/generate-type-forwarders/Program.cs:376]: `' '`

**Review by spouliot — APPROVED** [review]
minor

**Review by tj-devel709 — APPROVED** [review]

**Review by spouliot — APPROVED** [review]

**Comment by rolfbjarne** [comment]
Test failure is unrelated (https://github.com/xamarin/xamarin-macios/issues/10236)

**Pull request merged**